client/src/main/java/org/eclipse/hono/client/SendMessageSampler.java

```diff
@@ -0,0 +1,141 @@
+
+package org.eclipse.hono.client;
+
+/**
+ * An interface for sampling <em>send message</em> operations.
+ */
+public interface SendMessageSampler {
+
+    /**
+     * A factory for creating samplers.
+     */
+    interface Factory {
+
+        /**
+         * Create a new sampler.
+         *
+         * @param messageType The message type to create a sampler for.
+         * @return A new sampler.
+         */
+        SendMessageSampler create(String messageType);
+
+    }
+
+    /**
+     * Get a default, no-op implementation.
+     *
+     * @return A no-op implementation, never returns {@code null}.
+     */
+    static SendMessageSampler.Factory noop() {
+        return Noop.FACTORY;
+    }
+
+    /**
+     * Get a default, no-op implementation.
+     *
+     * @param messageType The message type to create the instance for.
+     * @return A no-op implementation, never returns {@code null}.
+     */
+    static SendMessageSampler noop(final String messageType) {
+        return noop().create(messageType);
+    }
+
+    /**
+     * A default no-op implementations.
+     */
+    class Noop {
+
+        private static final Sample SAMPLE = new Sample() {
+
+            @Override
+            public void accepted() {
+            }
+
+            @Override
+            public void rejected(final int errorCode) {
+            }
+
+            @Override
+            public void timeout() {
+            }
+
+        };
+
+        private static final SendMessageSampler SAMPLER = new SendMessageSampler() {
+
+            @Override
+            public Sample start(final String tenantId) {
+                return SAMPLE;
+            }
+
+            @Override
+            public void queueFull(final String tenantId) {
+            }
+
+        };
+
+        private static final Factory FACTORY = new Factory() {
+
+            @Override
+            public SendMessageSampler create(final String messageType) {
+                return SAMPLER;
+            }
+
+        };
+
+        private Noop() {
+        }
+
+    }
+
+    /**
+     * An active sample instance.
+     */
+    interface Sample {
+
+        /**
+         * Call when the message was accepted by the remote peer.
+         */
+        void accepted();
+
+        /**
+         * Call when the message was rejected by the remote peer.
+         *
+         * @param errorCode The rejection code. Expected to be a valid HTTP error code.
+         */
+        void rejected(int errorCode);
+
+        /**
+         * Call when the operation timed out.
+         */
+        void timeout();
+
+    }
+
+    /**
+     * Start operation.
+     *
+     * @param tenantId The tenant ID to sample for.
+     * @return A sample instance.
+     */
+    Sample start(String tenantId);
+
+    /**
+     * Record a case of "queue full".
+     * @param tenantId The tenant ID to sample for.
+     */
+    void queueFull(String tenantId);
```

Comment:
does this represent the case when the client is out of credit?

Reply:
I think so. It is derived from the fact that the `sendQueueFull()` method returns `true`, thus the name.